### PR TITLE
add debug log for memcache instantiation

### PR DESCRIPTION
**Pull request by butonic**
We should also think about increasing the log level when no cache is available to warning? cc @DeepDiver1975 @icewind1991

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
this will flood the logs in case we increase the log level to warning. We are elaborating if it might be better to explicitly require the admin to define the cache mechanisms in the config.

Especially if there are multiple available this will be necessary.

So :+1: for debug messages for now

**Comment by butonic** [comment]
@DeepDiver1975 I would use warning only for the else case where no cache is available. We should push our users to install at least APCu. That is as easy as one apt-get. And hosters usually have a memory cache installed.

Nevertheless, too many log entries ARE annoying and I feel more and more pushed to write a health app that shows some stats in the admin page ... like number of users, quota, which cache is used ... orphaned shares, dangling files in the filecache ... stuff like that.

**Comment by MorrisJobke** [comment]
:+1:

**Comment by MorrisJobke** [comment]
This creates 7 log entries for the files app opening only. Can we somehow cache these instances?

**Comment by butonic** [comment]
The idea of the debug log is to know what is going on. If we want to only see debug logging for certain apps then we need to make our logging more configurable. The current all or nothing situation makes certain situations undebuggable. I basically find myself chaining tail and several greps together to limit the output to what I need. Saying "let us write less debug messages *in general*" is the wrong path. Adding finer granularity like per app, or per class (a new context param) would be a start, IMNSHO.

**Comment by MorrisJobke** [comment]
@butonic sure :+1:

File: lib/private/memcache/factory.php [hunk 1]
```diff
@@ -31,17 +31,24 @@ public function __construct($globalPrefix) {
 	 */
 	function create($prefix = '') {
 		$prefix = $this->globalPrefix . '/' . $prefix;
+		$logger = \OC::$server->getLogger();
```
Comment (nickvergessen) [inline lib/private/memcache/factory.php]: Add to constructor?

File: lib/private/memcache/factory.php [hunk 2]
```diff
@@ -63,13 +70,18 @@ public function isAvailable() {
 	 */
 	public function createLowLatency($prefix = '') {
 		$prefix = $this->globalPrefix . '/' . $prefix;
+		$logger = \OC::$server->getLogger();
```
Comment (nickvergessen) [inline lib/private/memcache/factory.php]: same here

File: lib/private/memcache/factory.php [hunk 3]
```diff
@@ -31,17 +31,24 @@ public function __construct($globalPrefix) {
...
+			$logger->debug('creating APC instance', array('app'=>'memcache'));
 			return new APC($prefix);
 		} elseif (Redis::isAvailable()) {
+			$logger->debug('creating redis instance', array('app'=>'memcache'));
 			return new Redis($prefix);
 		} elseif (Memcached::isAvailable()) {
+			$logger->debug('creating memcached instance', array('app'=>'memcache'));
 			return new Memcached($prefix);
 		} else {
+			$logger->debug('no cache available instance', array('app'=>'memcache'));
```
Comment (nickvergessen) [inline lib/private/memcache/factory.php]: Remove `instance` here?

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
reassigning to 8.1 as this was merged against master - not critical from my pov - no need to backport this to stable8

**Pull request merged**